src/Tochka.JsonRpc.Client/Models/JsonRpcCallContext.cs

```diff
@@ -0,0 +1,253 @@
+        {
+            throw new InvalidOperationException("Can not add batch call when single call is present");
+        }
+
+        BatchCall = batchCall;
+        ExpectedBatchResponseCount = batchCall.Count(static x => x is UntypedRequest);
+    }
+
+    public void WithHttpResponse(HttpResponseMessage httpResponseMessage)
+    {
+        HttpResponseInfo = $"{httpResponseMessage}";
+        if (httpResponseMessage.StatusCode != HttpStatusCode.OK)
+        {
+            throw new JsonRpcException("Expected HTTP code 200", this);
+        }
+    }
+
+    public void WithHttpContent(HttpContent httpContent, string httpContentString)
+    {
+        HttpContentInfo = GetStringWithLimit(httpContentString);
+        if (httpContent == null)
+        {
+            throw new JsonRpcException("Response content is null", this);
+        }
+
+        var contentLength = httpContent.Headers.ContentLength;
+        if (contentLength is null or 0)
+        {
+            throw new JsonRpcException($"Bad Content-Length [{contentLength}]", this);
+        }
+    }
+
+    public void WithSingleResponse(IResponse singleResponse)
+    {
+        if (BatchResponse != null)
+        {
+            throw new InvalidOperationException("Can not add single response when batch response is present");
+        }
+
+        SingleResponse = singleResponse;
+        if (singleResponse == null)
+        {
+            throw new JsonRpcException("Could not parse body as JSON Rpc response", this);
+        }
+
+        if (singleResponse.Jsonrpc != JsonRpcConstants.Version)
+        {
+            throw new JsonRpcException($"JSON Rpc response version is invalid: [{singleResponse.Jsonrpc}], expected [{JsonRpcConstants.Version}]", this);
+        }
+
+        if (BatchCall != null)
+        {
+            return;
+        }
+
+        if (SingleCall is not UntypedRequest request)
+        {
+            throw new JsonRpcException("Received response but call was not request or batch", this);
+        }
+
+        // from specification:
+        // "If there was an error in detecting the id in the Request object (e.g. Parse error/Invalid Request), it MUST be Null."
+        if (SingleResponse is UntypedErrorResponse { Id: NullRpcId })
+        {
+            return;
+        }
+
+        if (!singleResponse.Id.Equals(request.Id))
+        {
+            throw new JsonRpcException($"JSON Rpc response id is invalid: [{singleResponse.Id}], expected [{request.Id}] or [null]", this);
+        }
+    }
+
+    public void WithBatchResponse(List<IResponse> batchResponse)
+    {
+        if (SingleResponse != null)
+        {
+            throw new InvalidOperationException("Can not add batch response when single response is present");
+        }
+
+        if (ExpectedBatchResponseCount == 0)
+        {
+            throw new InvalidOperationException("Can not add batch response when no response is expected");
+        }
+
+        BatchResponse = batchResponse;
```

Comment:
to bottom of method 

assignation gets lost here

Reply:
✅